include/mata/utils/partition-relation-pair.hh

```diff
@@ -0,0 +1,1449 @@
+/** @file partition_relation_pair.hh
+ *  @brief Definition of a partition, extendable square matrix
+ *  and partition-relation pair.
+ *
+ *  This file contains definitions of partition, extendable square
+ *  matrix, partition-relation pair and operations which allow us
+ *  to manipulate with them.
+ *
+ *  Description:
+ *
+ *  A partition-relation pair is a tuple (P, Rel). It is an efficient
+ *  representation of a preorder/quasiorder R, which is a reflexive and
+ *  transitive binary relation.
+ *  In this context, we consider a carrier set S which contains all
+ *  natural numbers from 0 to |S|-1. These numbers are called states.
+ *  P is a partition of S which corresponds to an equivalence relation
+ *  induced by the preorder R.
+ *  Rel is a partial order over P.
+ *  Thus, (P, Rel) corresponds to a preorder relation R over states S.
+ *  
+ *  This file provides implementation of a partition P and defines the
+ *  ExtendableSquareMatrix structure which can be used to represent 
+ *  the binary relation Rel. 
+ *  These structures can be combined to represent the preorder R.
+ *
+ *  @author Tomáš Kocourek
+ */
+
+#ifndef _PARTITION_RELATION_PAIR_HH_
+#define _PARTITION_RELATION_PAIR_HH_
+
+#include <iostream>
+#include <vector>
+#include <unordered_map>
+#include <cassert>
+
+namespace mata::utils {
+
+/************************************************************************
+*
+*
+*
+*                              PARTITION
+*                   
+*
+*
+*************************************************************************/
+
+
+using State = unsigned long;
+using StateBlock = std::vector<State>;
+using StateBlocks = std::vector<StateBlock>;
+
+using BlockItem = struct BlockItem {State state; size_t blockIdx;};
+using Block = struct Block {size_t nodeIdx;};
+using Node = struct Node {size_t first; size_t last;};
+
+using BlockItems = std::vector<BlockItem>;
+using Blocks = std::vector<Block>;
+using Nodes = std::vector<Node>;
+
+using SplitPair = struct SplitPair 
+                  {size_t former; size_t created; size_t oldNodeIdx;};
+
+/**
+ * Partition
+ * 
+ * @brief Partition of a set of states
+ *
+ * This data structure provides a partition of a set of states S. In this
+ * context, the term 'state' refers to any natural number from the
+ * interval <0, |S|-1>.
+ *
+ * STATES:
+ * This representation works with the vector of indices 'm_states'
+ * with the constant size |S|. Each state is represented by an index
+ * to that vector so we can refer to a state in constant time using
+ * m_states[state].
+ *
+ * BLOCK ITEMS:
+ * The memory cell m_states[state] contains a corresponding index to the 
+ * 'm_blockItems' vector. The vector 'm_blockItems' has the constant size |S|.
```

Comment:
In Mata, we agreed to use `member_name_` convention (trailing `_`) to denote member variables/functions. So instead of `m_blockItems`, the member should be called `block_items_`.

We also use `snake_case` instead of `lowerCamelCase`.

Reply:
Thanks for your suggestion! I have committed these changes:

- private attributes `m_varName` renamed to `var_name_` within all classes
- `lowerCamelCase` replaced by `snake_case` in context of 
    - variables
    - functions (which are neither constructors nor destructors)
- `UpperCamelCase` was kept in context of
    - class/structure names
    - aliases created using `using`
    - constructors and destructors
    - elements of the type `enum`
    
Is this ok? If so, the suggestion could be marked as resolved.

Reply:
This is amazing. Great work.